lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs

```diff
@@ -564,35 +532,53 @@ extendAddressPool !ix !pool
   where
     edge = Map.size (indexedKeys pool)
     isOnEdge = edge - fromEnum ix <= fromEnum (gap pool)
-    next = if ix == maxBound then mempty else nextAddresses @n
-        (accountPubKey pool)
+    next = if ix == maxBound then mempty else nextAddresses @n @c
+        (context pool)
         (gap pool)
-        (role @c)
         (succ ix)
 
 -- | Compute the pool extension from a starting index
 nextAddresses
-    :: forall (n :: NetworkDiscriminant) k.
+    :: forall (n :: NetworkDiscriminant) (c :: Role) k.
         ( MkKeyFingerprint k (Proxy n, k 'AddressK XPub)
+        , MkKeyFingerprint k Address
         , SoftDerivation k
+        , Typeable c
+        , Typeable n
         )
-    => k 'AccountK XPub
+    => ParentContext c k
     -> AddressPoolGap
-    -> Role
     -> Index 'Soft 'AddressK
     -> Map (KeyFingerprint "payment" k) (Index 'Soft 'AddressK, AddressState)
-nextAddresses !key (AddressPoolGap !g) !cc !fromIx =
+nextAddresses !ctx (AddressPoolGap !g) !fromIx =
     [fromIx .. min maxBound toIx]
-        & map (\ix -> (newPaymentKey ix, (ix, Unused)))
+        & map (\ix -> (mkPaymentKey ix ctx, (ix, Unused)))
         & Map.fromList
   where
     toIx = invariant
         "nextAddresses: toIx should be greater than fromIx"
         (toEnum $ fromEnum fromIx + fromEnum g - 1)
         (>= fromIx)
-    newPaymentKey = unsafePaymentKeyFingerprint @k
-        . (Proxy @n,)
-        . deriveAddressPublicKey key cc
+
+    mkPaymentKey ix = \case
+        ParentContextUtxoExternal acct ->
+            mkPaymentKeyFromAccXPub acct
+        ParentContextUtxoInternal acct ->
+            mkPaymentKeyFromAccXPub acct
+        ParentContextMutableAccount acct ->
+            mkPaymentKeyFromAccXPub acct  -- think about whether it makes sense
```

Comment:
Spoiler: I don't think it does :sweat_smile:

Reply:
:smile:  removed in https://github.com/input-output-hk/cardano-wallet/pull/2508/commits/37e818cc20a5248c85f8645e712295541941e614